File: backend/app/models/learning.py
"""学习Agent数据模型"""
import uuid
from datetime import datetime, timezone
from sqlalchemy import String, Integer, Float, Text, DateTime, Boolean
from sqlalchemy.orm import Mapped, mapped_column

from app.db.connection import Base


def _utcnow() -> datetime:
    return datetime.now(timezone.utc)


def _uuid() -> str:
    return str(uuid.uuid4())


class HotNovelMeta(Base):
    __tablename__ = "hot_novel_meta"

    id: Mapped[str] = mapped_column(String(64), primary_key=True, default=_uuid)
    platform: Mapped[str] = mapped_column(String(32), nullable=False)  # fanqie / qidian
    source_book_id: Mapped[str] = mapped_column(String(64), default="")  # 平台侧 book_id
    title: Mapped[str] = mapped_column(String(256), nullable=False)
    author: Mapped[str] = mapped_column(String(128), default="")
    genre: Mapped[str] = mapped_column(String(64), default="")
    tags: Mapped[str] = mapped_column(Text, default="[]")  # JSON
    word_count: Mapped[int] = mapped_column(Integer, default=0)
    chapter_count: Mapped[int] = mapped_column(Integer, default=0)
    rating: Mapped[float] = mapped_column(Float, nullable=True)
    read_count: Mapped[int] = mapped_column(Integer, default=0)  # 累计阅读人数
    bookshelf_count: Mapped[int] = mapped_column(Integer, default=0)  # 收藏人数
    created_at_source: Mapped[str] = mapped_column(String(64), default="")  # 平台侧创建时间 ISO
    synopsis: Mapped[str] = mapped_column(Text, default="")
    cover_url: Mapped[str] = mapped_column(String(512), default="")
    rank_info: Mapped[str] = mapped_column(Text, default="{}")  # JSON: {rank_type: position}
    source_url: Mapped[str] = mapped_column(String(512), default="")
    status: Mapped[str] = mapped_column(String(16), default="meta")  # meta / crawling / done / failed
    crawled_at: Mapped[datetime] = mapped_column(DateTime, default=_utcnow)


class HotNovelChapter(Base):
    __tablename__ = "hot_novel_chapters"

    id: Mapped[str] = mapped_column(String(64), primary_key=True, default=_uuid)
    novel_id: Mapped[str] = mapped_column(String(64), nullable=False, index=True)
    source_chapter_id: Mapped[str] = mapped_column(String(64), default="")
    chapter_number: Mapped[int] = mapped_column(Integer, default=0)
    title: Mapped[str] = mapped_column(String(256), default="")
    content: Mapped[str] = mapped_column(Text, default="")
    word_count: Mapped[int] = mapped_column(Integer, default=0)
    crawled_at: Mapped[datetime] = mapped_column(DateTime, default=_utcnow)


class KnowledgeEntry(Base):
    __tablename__ = "knowledge_entries"

    id: Mapped[str] = mapped_column(String(64), primary_key=True, default=_uuid)
    category: Mapped[str] = mapped_column(String(64), nullable=False)
    # opening_pattern / thrill_distribution / character_template / dialogue_style
    # writing_style / foreshadow_technique / genre_formula
    title: Mapped[str] = mapped_column(String(256), nullable=False)
    content: Mapped[str] = mapped_column(Text, nullable=False)
    source_novel_id: Mapped[str] = mapped_column(String(64), nullable=True)
    source_file: Mapped[str] = mapped_column(String(512), default="")  # 教程导入时的文件路径
    tags: Mapped[str] = mapped_column(Text, default="[]")  # JSON
    quality_score: Mapped[float] = mapped_column(Float, default=0.0)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=_utcnow)
    expires_at: Mapped[datetime] = mapped_column(DateTime, nullable=True)


class OptimizationLog(Base):
    __tablename__ = "optimization_logs"

    id: Mapped[str] = mapped_column(String(64), primary_key=True, default=_uuid)
    target: Mapped[str] = mapped_column(String(64), nullable=False)  # prompt / workflow / model_config
    description: Mapped[str] = mapped_column(Text, default="")
    before_snapshot: Mapped[str] = mapped_column(Text, default="{}")  # JSON
    after_snapshot: Mapped[str] = mapped_column(Text, default="{}")  # JSON
    improvement_score: Mapped[float] = mapped_column(Float, nullable=True)
    applied: Mapped[bool] = mapped_column(Boolean, default=False)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=_utcnow)

File: frontend/src/api/learning.ts
import apiClient from './client'

/* ── Types ───────────────────────────────────────────────── */

export interface LearningStats {
  knowledge_count: number
  hot_novel_count: number
  chapter_count: number
  opt_log_count: number
  crawling_count: number
  done_count: number
  last_crawl_at: string | null
}

export interface KnowledgeEntry {
  id: string
  category: string
  title: string
  content: Record<string, any>
  source_novel_id: string | null
  source_file: string
  tags: string[]
  quality_score: number
  created_at: string | null
  expires_at?: string | null
}

export interface TutorialScanResult {
  total: number
  by_category: Record<string, number>
  files: { path: string; relative: string; name: string; ext: string; size: number; category: string; folder: string }[]
}

export interface TutorialImportResult {
  status: string
  message: string
}

export interface HotNovel {
  id: string
  platform: string
  source_book_id: string
  title: string
  author: string
  genre: string
  tags: string[]
  word_count: number
  chapter_count: number
  rating: number
  synopsis: string
  cover_url: string
  rank_info: Record<string, any>
  source_url: string
  status: string   // meta | crawling | done | failed
  crawled_at: string | null
}

export interface HotNovelChapter {
  id: string
  chapter_number: number
  title: string
  word_count: number
}

export interface ChapterContent extends HotNovelChapter {
  content: string
}

export interface OptimizationLog {
  id: string
  target: string
  description: string
  before_snapshot: Record<string, any>
  after_snapshot: Record<string, any>
  improvement_score: number
  applied: boolean
  created_at: string | null
}

export interface TriggerResult {
  status: string
  message: string
}

/* ── API ─────────────────────────────────────────────────── */

export const learningApi = {
  /* 统计 */
  getStats: () =>
    apiClient.get<LearningStats>('/learning/stats').then(r => r.data),

  /* 知识库 */
  listKnowledge: (category?: string, limit = 50) =>
    apiClient.get<KnowledgeEntry[]>('/learning/knowledge', { params: { category, limit } }).then(r => r.data),

  getKnowledge: (id: string) =>
    apiClient.get<KnowledgeEntry>(`/learning/knowledge/${id}`).then(r => r.data),

  deleteKnowledge: (id: string) =>
    apiClient.delete(`/learning/knowledge/${id}`).then(r => r.data),

  /* 热门小说 */
  listHotNovels: (platform?: string, limit = 50) =>
    apiClient.get<HotNovel[]>('/learning/hot-novels', { params: { platform, limit } }).then(r => r.data),

  listNovelChapters: (novelId: string) =>
    apiClient.get<HotNovelChapter[]>(`/learning/hot-novels/${novelId}/chapters`).then(r => r.data),

  getChapterContent: (novelId: string, chapterId: string) =>
    apiClient.get<ChapterContent>(`/learning/hot-novels/${novelId}/chapters/${chapterId}`).then(r => r.data),

  /* 优化日志 */
  listOptLogs: (limit = 30) =>
    apiClient.get<OptimizationLog[]>('/learning/optimization-logs', { params: { limit } }).then(r => r.data),

  applyOptLog: (logId: string) =>
    apiClient.post<{ ok: boolean; log_id: string }>(`/learning/optimization-logs/${logId}/apply`).then(r => r.data),

  /* 手动触发 */
  triggerCrawl: () =>
    apiClient.post<TriggerResult>('/learning/trigger-crawl').then(r => r.data),

  triggerLearn: () =>
    apiClient.post<TriggerResult>('/learning/trigger-learn').then(r => r.data),

  triggerOptimize: () =>
    apiClient.post<TriggerResult>('/learning/trigger-optimize').then(r => r.data),

  triggerCoverDownload: () =>
    apiClient.post<TriggerResult>('/learning/trigger-cover-download').then(r => r.data),

  /* 教程导入 */
  scanTutorials: (baseDir: string) =>
    apiClient.post<TutorialScanResult>('/learning/tutorial/scan', null, { params: { base_dir: baseDir } }).then(r => r.data),

  importTutorials: (baseDir: string, useLlm = true, maxFiles = 100) =>
    apiClient.post<TutorialImportResult>('/learning/tutorial/import', null, { params: { base_dir: baseDir, use_llm: useLlm, max_files: maxFiles } }).then(r => r.data),

  importSingleTutorial: (filePath: string, category?: string, useLlm = true) =>
    apiClient.post('/learning/tutorial/import-file', null, { params: { file_path: filePath, category, use_llm: useLlm } }).then(r => r.data),

  getActivityLog: (since = 0) =>
    apiClient.get<{ total: number; logs: { ts: string; level: string; msg: string }[] }>('/learning/activity-log', { params: { since } }).then(r => r.data),

  /* 番茄登录状态（复用发布中心） */
  fanqieLoginStatus: () =>
    apiClient.get<{ logged_in: boolean; cookie_count?: number; msg: string }>('/learning/fanqie/login-status').then(r => r.data),
}
